include/invert_quda.h

```diff
@@ -657,42 +657,67 @@ namespace quda {
     virtual bool hermitian() { return true; } /** CG is only for Hermitian systems */
   };
 
-
-  class CG3 : public Solver {
+  class CGNE : public CG
+  {
 
   private:
-    // pointers to fields to avoid multiple creation overhead
-    ColorSpinorField *yp, *rp, *tmpp, *ArSp, *rSp, *xSp, *xS_oldp, *tmpSp, *rS_oldp, *tmp2Sp;
+    DiracMMdag mmdag;
+    DiracMMdag mmdagSloppy;
+    DiracMMdag mmdagPrecon;
+    ColorSpinorField *xp;
+    ColorSpinorField *yp;
     bool init;
 
   public:
-    CG3(const DiracMatrix &mat, const DiracMatrix &matSloppy, SolverParam &param, TimeProfile &profile);
-    virtual ~CG3();
+    CGNE(const DiracMatrix &mat, const DiracMatrix &matSloppy, const DiracMatrix &matPrecon, SolverParam &param,
+         TimeProfile &profile);
+    virtual ~CGNE();
 
     void operator()(ColorSpinorField &out, ColorSpinorField &in);
 
-    virtual bool hermitian() { return true; } /** CG is only for Hermitian systems */
+    virtual bool hermitian() { return false; } /** CGNE is for any system */
   };
 
+  class CGNR : public CG
+  {
+
+  private:
+    DiracMdagM mdagm;
+    DiracMdagM mdagmSloppy;
+    DiracMdagM mdagmPrecon;
+    ColorSpinorField *bp;
+    bool init;
+
+  public:
+    CGNR(const DiracMatrix &mat, const DiracMatrix &matSloppy, const DiracMatrix &matPrecon, SolverParam &param,
+         TimeProfile &profile);
+    virtual ~CGNR();
 
-  class CG3NE : public Solver {
+    void operator()(ColorSpinorField &out, ColorSpinorField &in);
+
+    virtual bool hermitian() { return false; } /** CGNR is for any system */
+  };
+
+  class CG3 : public Solver
```

Comment:
Orthogonal question---is there any reason we still have CG3? Is anyone actually using it?

Reply:
@sbacchio ?